tests/executequery_test.go

```diff
@@ -99,10 +174,1137 @@ func TestExecuteQuery_SingleSimpleRow(t *testing.T) {
+		&executeQueryAction{
+			response:    partialResultSet("token", bytesVal([]byte("foo"))), // Only 1 value provided with token
+			endOfStream: false,
+		})
+	// 2. Build the request to test proxy
+	req := testproxypb.ExecuteQueryRequest{
+		ClientId: t.Name(),
+		Request: &btpb.ExecuteQueryRequest{
+			InstanceName: instanceName,
+			Query:        "SELECT * FROM table",
+		},
+	}
+	// 3. Perform the operation via test proxy
+	res := doExecuteQueryOp(t, server, &req, nil)
+	// 4. Verify the operation fails with an appropriate error about incomplete data
+	assert.Equal(t, int32(codes.Internal), res.GetStatus().GetCode())
+	assertErrorIn(t, res, []string{
+		// Java error message
+		"Incomplete row received",
+		// Python error message
+		"InvalidExecuteQueryResponse",
+	})
+}
+
+// Tests that the operation fails if the response stream ends with an incomplete row, even after complete rows
+func TestExecuteQuery_FailsOnNotEnoughDataWithCompleteRows(t *testing.T) {
+	// 1. Instantiate the mock server to return a complete row, then partial row data
+	server := initMockServer(t)
+	server.PrepareQueryFn = mockPrepareQueryFn(nil,
+		&prepareQueryAction{
+			response: prepareResponse([]byte("foo"), md(
+				column("foo", bytesType()),
+				column("bar", strType()), // Expects 2 values per row
+			)),
+		},
+	)
+	server.ExecuteQueryFn = mockExecuteQueryFn(nil,
+		&executeQueryAction{
+			// One complete row (2 values) + start of next row (1 value)
+			response:    partialResultSet("token", bytesVal([]byte("foo")), strVal("s"), bytesVal([]byte("bar"))),
+			endOfStream: false,
+		})
+	// 2. Build the request to test proxy
+	req := testproxypb.ExecuteQueryRequest{
+		ClientId: t.Name(),
+		Request: &btpb.ExecuteQueryRequest{
+			InstanceName: instanceName,
+			Query:        "SELECT * FROM table",
+		},
+	}
+	// 3. Perform the operation via test proxy
+	res := doExecuteQueryOp(t, server, &req, nil)
+	// 4. Verify the operation fails with an appropriate error about incomplete data
+	assert.Equal(t, int32(codes.Internal), res.GetStatus().GetCode())
+	assertErrorIn(t, res, []string{
+		// Java error message
+		"Incomplete row received",
+		// Python error message
+		"InvalidExecuteQueryResponse",
+	})
+}
+
+// Tests that the operation fails if received data type mismatches the metadata type
+func TestExecuteQuery_FailsOnTypeMismatch(t *testing.T) {
+	// 1. Instantiate the mock server to return data with a type mismatch
+	server := initMockServer(t)
+	server.PrepareQueryFn = mockPrepareQueryFn(nil,
+		&prepareQueryAction{
+			response: prepareResponse([]byte("foo"), md(
+				column("foo", bytesType()),
+				column("bar", strType()), // Expects string for 2nd column
+			)),
+		},
+	)
+	server.ExecuteQueryFn = mockExecuteQueryFn(nil,
+		&executeQueryAction{
+			response:    partialResultSet("token", bytesVal([]byte("foo")), strVal("s"), bytesVal([]byte("bar")), intVal(42)), // Sends int where string expected
+			endOfStream: false,
```

Comment:
should endOfStream be true? just so that the only invalid thing here is the wrong type. Same for the test below

Reply:
good point. Done